### PR TITLE
feat(web): fileserver implementation with new dependencies

### PR DESCRIPTION
**Pull request by Phillip9587**
## Current Behavior
@nrwl/web file-server executor uses `node-watch` and `http-server`. 

## Expected Behavior
`node-watch` can replaced with the much more popular `chokidar` package and `http-server` with the node http/https server. I also used [`connect`](https://www.npmjs.com/package/connect) because it is very lightweight and allows the use of middlewares. For serving the static files i used the popular [`serve-static`](https://www.npmjs.com/package/serve-static) middleware.

It would also be possible to create a custom proxy implementation with [http-proxy](https://www.npmjs.com/package/http-proxy) but i decided to use the popular [`http-proxy-middleware`](https://www.npmjs.com/package/http-proxy-middleware).

## Related Issue(s)
#5305

Fixes #

@juristr Would you mind reviewing the changes?

### CODE REVIEW

File: packages/web/src/builders/file-server/file-server.impl.ts [hunk 1]
```diff
@@ -125,38 +95,53 @@ export default async function* fileServerExecutor(
...
 
-  const outputPath = getBuildTargetOutputPath(opts, context);
-  const args = getHttpServerArgs(opts);
+  const outputPath = getBuildTargetOutputPath(options, context);
+
+  const app = connect();
+
+  app.use(serveStatic(outputPath, { fallthrough: true }));
+
+  if (options.proxyUrl)
```
Comment (juristr) [inline packages/web/src/builders/file-server/file-server.impl.ts:104]: I prefer to have `{...}` around these if statements to have it consistent with the rest of the codebase and make it more readable :)

**Review by juristr — CHANGES_REQUESTED** [review]

**Comment by juristr** [comment]
@Phillip9587 Thx for the Chokidar change, that was planned, we just didn't came around to add it yet. So thx for that :)

Related to the HTTP server change, I've been discussing with @puku0x [here](https://github.com/nrwl/nx/pull/5305#issuecomment-817941438) about possible alternatives to support the SPA scenario. Right now (and I believe also with your change), if you navigate to some route, and refresh, it fails because it doesn't fallback to the `index.html` and delegate to the client-side routing.

I believe it should be possible to handle this by adding some middleware to the current implementation as well, right?

**Comment by Phillip9587** [comment]
@juristr Yes it should be possible by adding a middleware before the proxy middleware i guess.

**Comment by Phillip9587** [comment]
@juristr But by redirecting all 404 to the `index.html` the request can't be proxied anymore?

**Comment by Phillip9587** [comment]
@juristr Should we introduce a spa option to the executor schema which should be mutually exclusive with the [proxyUrl](https://nx.dev/latest/angular/web/file-server#proxyurl) option. Or SPA Fallback to `index.html` could be the default behaviour as long as someone has no proxyUrl option specified?

File: packages/web/src/builders/file-server/file-server.impl.ts (hunk 1 above)
Comment (Phillip9587) [reply packages/web/src/builders/file-server/file-server.impl.ts]: Done

**Comment by Phillip9587** [comment]
@juristr @puku0x  I Implemented a basic fallback to `index.html` if the proxyUrl option is not set and the `index.html` file exists in the output directory.

**Comment by juristr** [comment]
cool thx, won't make it today, but will check it out tomorrow 👍

**Comment by Phillip9587** [comment]
@juristr Are there any changes to make?

**Comment by juristr** [comment]
@Phillip9587 we had some internal discussions exactly about improvements on the file server, so I did hold before continuing here. I'll be back with some feedback asap

**Comment by Phillip9587** [comment]
@juristr Thanks

File: packages/web/src/builders/file-server/file-server.impl.ts [hunk 2]
```diff
@@ -125,38 +102,78 @@ export default async function* fileServerExecutor(
...
-      process.stdout.write(chunk);
+  const app = connect();
+
+  app.use(serveStatic(outputPath, { fallthrough: true }));
+
+  if (options.proxyUrl) {
+    app.use(createProxyMiddleware({ target: options.proxyUrl }));
+  } else {
+    // SPA Fallback to index.html
+    app.use((_req, res, next) => {
```
Comment (juristr) [inline packages/web/src/builders/file-server/file-server.impl.ts:115]: 🤔 I think we shouldn't just do the index.html fallback when there is no proxy. With the webpack dev server for example, you can easily have a proxy configuration, but still have the fallback part.

We might want to look at the configured proxy URLs. Maybe we can reuse the webpack devserver logic or middleware that is being used there for providing the same behavior.

Comment (Phillip9587) [reply packages/web/src/builders/file-server/file-server.impl.ts:115]: @juristr I will take a look at this this week. I will go through the sourcecode of the webpack-dev-middleware package maybe we can use it directly or just parts of it.

**Comment by Phillip9587** [comment]
@juristr I'm sorry i currently don't have time to look into the webpack dev-server middleware. But i can update the pull request to only make the change to chokidar so that at least something was improved...

**Comment by juristr** [comment]
> @juristr I'm sorry i currently don't have time to look into the webpack dev-server middleware. But i can update the pull request to only make the change to chokidar so that at least something was improved...

sure, that sounds good. we can merge that right away. Some of us can take over your other improvements and work from there. So if you extract the chokidar piece and remove it from this PR would be awesome.

**Pull request closed without merging**